### PR TITLE
i#3048 func-trace: check redzone to see if write out trace data after appending markers

### LINKED ISSUE #3048: Record key function traces in drcachesim
Add mechanism to collect traces for specified/key functions in drcachesim. The traces should integrate with memory accesses traces.
The new traces should include following information for each function:
1. function invocation arguments
2. function return address
3. function return value

### PR DESCRIPTION
**Pull request by Louis-Ye**
The append_marker_seg_base, used by func_trace module, does not write out trace data to buffer after appending the marker entries, which (though very unlikely) could overflow the buffer if the markers are appended before a large BB.

Fixes #3048

### CODE REVIEW

File: clients/drcachesim/tracer/tracer.cpp [hunk 1]
```diff
@@ -540,6 +540,8 @@ append_marker_seg_base(void *drcontext, trace_marker_type_t marker, uintptr_t va
         return; /* This thread was filtered out. */
     BUF_PTR(data->seg_base) +=
         instru->append_marker(BUF_PTR(data->seg_base), marker, value);
+    if (file_ops_func.handoff_buf == NULL)
+        memtrace(drcontext, false);
```
Comment (derekbruening) [inline clients/drcachesim/tracer/tracer.cpp]: I don't think we want to call this every time: we're going to end up dumping a single-entry buffer over and over for the retaddr and func arg markers.  We need to add a check for whether the buffer has hit the redzone (in C++ here -- today we only have the check in the generated instru).

Comment (Louis-Ye) [reply clients/drcachesim/tracer/tracer.cpp]: Done.

File: clients/drcachesim/tracer/tracer.cpp [hunk 2]
```diff
@@ -540,6 +540,13 @@ append_marker_seg_base(void *drcontext, trace_marker_type_t marker, uintptr_t va
         return; /* This thread was filtered out. */
     BUF_PTR(data->seg_base) +=
         instru->append_marker(BUF_PTR(data->seg_base), marker, value);
+    if (file_ops_func.handoff_buf == NULL) {
+        byte *buf_ptr = BUF_PTR(data->seg_base);
+        byte *redzone = data->buf_base + trace_buf_size;
+        if (buf_ptr > redzone) {
+            memtrace(drcontext, false);
```
Comment (derekbruening) [inline clients/drcachesim/tracer/tracer.cpp]: Are there any invariants promised by the tracer, or assumptions made by function marker readers, that no thread switch occurs in the middle of the sequence of function entry markers (id, retaddr, args)?  Are you putting a burden on readers to either split by thread up front or have to cache markers seen for resumption of the current thread?  I wonder if we should make this guarantee to simplify readers (and remove this buffer dump)?

Could you elaborate your logic behind worrying about hitting the buffer end here?  The func_pre_hook will be on function entry, which should be at the bb start.  Adding the extra marker entries there could cause the memrefs later in the bb to go further into the buffer than they would without function tracing: isn't that just as likely if not more likely than that the markers themselves, and thus this check, would cross over?  The redzone is supposed to allow a lot of slop and let us overflow and catch it afterward at the end-of-bb check.  Will adding a handful of markers really change the risk of going off the end of the redzone?

Comment (Louis-Ye) [reply clients/drcachesim/tracer/tracer.cpp]: Summary for the offline discussion:
* memtrace (or thread switch stamp) happening between consecutive of markers is bad, we should avoid it.
* risk of going off redzone still exists if markers appended before a large BB.
The approach here is to check the redzone only after appending a consecutive number of markers.

Done.

**Comment by derekbruening** [comment]
Travis failed: looks like a clang-format violation

**Pull request merged**